### PR TITLE
Implement PMGSNES

### PR DESCRIPTION
**Pull request by pbrubeck**
Here we introduce the class `PMGSNES`, which enables the use of MG FAS for nonlinear problems.
The key main features are:
- Patch to enable assembly with symmetric TensorFunctionSpaces.
- Patch to allow matrix free projection.
- Fix logic errors in `tensor_product_space_query()`
- Matrix-free injection of the state.
- Coarsening of boundary conditions.
- Built-in coarsening of VectorElement, TensorElement, and TensorProductElements with a single polynomial degree.
- Coarsening of user-provided quadrature degree that preserves the ratio of interpolation (e.g. GLL) to quadrature (e.g. GL) nodes.
- Fix a bug related to the BLAS library flag when compiling the grid transfer kernel.

### CODE REVIEW

**Comment by wence-** [comment]
Hi @pbrubeck from a cursory glance this looks like a nice feature. I don't have a lot of time in the next 2 weeks for reviews, but then stuff should ease up, so if I haven't looked by then, please give me a prod.

File: firedrake/preconditioners/pmg.py [hunk 1]
```diff
@@ -62,12 +62,42 @@ def coarsen_element(ele):
...
+
+        Must provide the same degree for tensor product elements.
+
+        :arg ele: a :class:`ufl.FiniteElement` to reconstruct.
+        :arg N: an integer degree.
+        """
+        if isinstance(ele, (VectorElement, TensorElement)):
+            cVe = [PMGBase.reconstruct_degree(K, N)
+                   for K in ele.sub_elements()]
+            return VectorElement(cVe[0], dim=len(cVe))
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:93]: Why `reconstruct_degree` on all subelements, when you only use the first one?

File: firedrake/preconditioners/pmg.py [hunk 2]
```diff
@@ -62,12 +62,42 @@ def coarsen_element(ele):
...
+        elif isinstance(ele, TensorProductElement):
+            (Ve, Qe) = ele.sub_elements()
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:95]: Do all `TensorProductElement`s only wrap two subelements?

File: firedrake/preconditioners/pmg.py [hunk 3]
```diff
@@ -103,44 +133,27 @@ def initialize(self, pc):
...
+        pdm.setSNESJacobian(_SNESContext.form_jacobian)
+        pdm.setSNESFunction(_SNESContext.form_function)
+        pdm.setKSPComputeOperators(_SNESContext.compute_operators)
+
         pdm.setOptionsPrefix(pc.getOptionsPrefix() + "pmg_")
         set_function_space(pdm, get_function_space(odm))
 
         parent = get_parent(odm)
         assert parent is not None
-        add_hook(parent, setup=partial(push_parent, pdm, parent), teardown=partial(pop_parent, pdm, parent),
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:111]: Was it necessary to change these lines (for lint etc)?

File: firedrake/preconditioners/pmg.py [hunk 4]
```diff
@@ -103,44 +133,27 @@ def initialize(self, pc):
...
+        add_hook(parent, setup=partial(push_parent, pdm, parent),
+                 teardown=partial(pop_parent, pdm, parent),
                  call_setup=True)
-        add_hook(parent, setup=partial(push_appctx, pdm, ctx), teardown=partial(pop_appctx, pdm, ctx),
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:113]: Same

File: firedrake/preconditioners/pmg.py [hunk 5]
```diff
@@ -154,17 +167,19 @@ def coarsen(self, fdm, comm):
...
+        Injection = prolongation_matrix_matfree(cV, fV, [], [])
+
+        def inject_state():
+            with cu.dat.vec_wo as xc, fu.dat.vec_ro as xf:
+                Injection.multTranspose(xf, xc)
 
         parent = get_parent(fdm)
         assert parent is not None
-        add_hook(parent, setup=partial(push_parent, cdm, parent), teardown=partial(pop_parent, cdm, parent),
+        add_hook(parent,
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:180]: Why break up into multiple lines?

File: firedrake/preconditioners/pmg.py [hunk 6]
```diff
@@ -179,20 +194,40 @@ def inject_state(interpolators):
...
+                    N, = set(N)
+                except TypeError:
+                    pass
+                return N
+
+        # Coarsen the quadrature degree
+        # such that the ratio of GL to GLL nodes is preserved
         fcp = fctx._problem.form_compiler_parameters
-        cproblem = firedrake.NonlinearVariationalProblem(cF, cu, cbcs, cJ,
+        if fcp is not None:
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:221]: I think it's more usual to specify the quadrature degree by passing options to the `Measure` object. I don't think this will catch that case. Is that correct?

File: firedrake/preconditioners/pmg.py [hunk 7]
```diff
@@ -303,18 +453,20 @@ def tensor_product_space_query(V):
     if isinstance(ele, firedrake.TensorProductElement):
         family = set(e.family() for e in ele.sub_elements())
         try:
-            variant, = set(e.variant() or "spectral" for e in ele.sub_elements())
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:306]: Can you comment on what the changes around here are doing?

File: firedrake/preconditioners/pmg.py [hunk 8]
```diff
@@ -529,8 +682,8 @@ def make_blas_kernels(self, Vf, Vc):
             return;
         }}
         """
-        self.prolong_kernel = op2.Kernel(prolong_code, "prolongation", ldargs=["-lpetsc"])
-        self.restrict_kernel = op2.Kernel(restrict_code, "restriction", ldargs=["-lpetsc"])
+        self.prolong_kernel = op2.Kernel(prolong_code, "prolongation", ldargs=["-lblas"])
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:685]: On supercomputers we usually get our BLAS from libraries other than `libblas.so`. Is there a way to figure out cleanly what BLAS we should be linking against here? I think the rationale for using PETSc was that it would be linked to the correct BLAS, but correct me if I'm wrong.

**Review by pefarrell — CHANGES_REQUESTED** [review]
Can you add some tests that show off the new functionality?

File: firedrake/preconditioners/pmg.py (hunk 3 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:111]: Not at all, it was just my obsesion to split lines at 77 characters.

File: firedrake/preconditioners/pmg.py (hunk 6 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:221]: In that case, whenever we specify the degree on the measure, could I just add a measure with the coarse degree in the replace dictionary?

File: firedrake/preconditioners/pmg.py [hunk 9]
```diff
@@ -62,12 +62,38 @@ def coarsen_element(ele):
...
+        :arg ele: a :class:`ufl.FiniteElement` to reconstruct.
+        :arg N: an integer degree.
+        """
+        if isinstance(ele, (VectorElement, TensorElement)):
+            sub = ele.sub_elements()
+            return VectorElement(PMGBase.reconstruct_degree(sub[0], N), dim=len(sub))
+        elif isinstance(ele, TensorProductElement):
+            return TensorProductElement(*(PMGBase.reconstruct_degree(sub, N) for sub in ele.sub_elements()))
+        else:
+            return ele.reconstruct(degree=N)
```
Comment (rckirby) [inline firedrake/preconditioners/pmg.py:97]: This work is interesting, and I've needed similar things for P1PC-type things in tensor product cells.  It seems like the reconstruction should/could be done at the UFL level instead?

File: firedrake/preconditioners/pmg.py (hunk 2 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py]: Good point, it should be extended to multiple subelements.

File: firedrake/preconditioners/pmg.py (hunk 7 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:306]: For one moment I thought there was a bug here, after checking twice, I think that it should be kept as it was before.

File: firedrake/preconditioners/pmg.py (hunk 8 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:674]: @djanekovic might have some useful input on this one...

Comment (djanekovic) [reply firedrake/preconditioners/pmg.py:672]: This was bug I introduced and Pablo fixed. It is maybe confusing why we are including `petscblaslapack.h` but we need to link to `libblas.so`. That header file implements just the name mangling logic in order to call correct BLAS/LAPACK function. Because of that, we need to link to the `libblas.so` in the end.
Patrick, yes, you are correct and this problem was solved just recently. Jack added `get_petsc_variables` function in his PR #1928 which enables getting a path to the BLAS/LAPACK provided to the PETSc configure.
To sum up, Pablo, to link with correct BLAS library you should implement the same logic Jack implemented here: https://github.com/firedrakeproject/firedrake/blob/master/firedrake/slate/slac/compiler.py#L69

File: firedrake/preconditioners/pmg.py (hunk 6 above)
Comment (pefarrell) [reply firedrake/preconditioners/pmg.py:210]: Does `ufl.replace' replace `Measure`s? I have never checked, but I wouldn't think so.

File: firedrake/preconditioners/pmg.py (hunk 7 above)
Comment (pefarrell) [reply firedrake/preconditioners/pmg.py:306]: Can  you just add a comment explaining what's going on? (Especially if you got confused yourself!)

File: firedrake/preconditioners/pmg.py (hunk 9 above)
Comment (pefarrell) [reply firedrake/preconditioners/pmg.py:97]: So `P1PC` is a subclass of `PMGPC` but with `coarsen_element` overridden. One solution is to get @pbrubeck to modify that the overridden `coarsen_element` appropriately in this PR or a later PR also. But I think what @rckirby is asking is if this knowledge should go back into UFL. @wence- , any thoughts?

**Comment by rckirby** [comment]
Specifically, coarsening of elements put into UFL and then used here rather
than done here.

File: firedrake/preconditioners/pmg.py [hunk 10]
```diff
@@ -529,8 +670,21 @@ def make_blas_kernels(self, Vf, Vc):
             return;
         }}
         """
-        self.prolong_kernel = op2.Kernel(prolong_code, "prolongation", ldargs=["-lpetsc"])
-        self.restrict_kernel = op2.Kernel(restrict_code, "restriction", ldargs=["-lpetsc"])
+        BLASLAPACK_LIB = None
+        BLASLAPACK_INCLUDE = None
+        if not complex_mode:
```
Comment (djanekovic) [inline firedrake/preconditioners/pmg.py:675]: I don't think you need this here since everything is complex safe.

File: firedrake/preconditioners/pmg.py (hunk 9 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:98]: I can easily extend this coarsening to `P1PC`. But a question arises: do we want the lowest order for a familty (e.g. degree = 0 in the case of DG) or stick to the naming and have always degree = 1?

**Comment by rckirby** [comment]
Unclear.  DG(0) is lowest order, but some DG methods require degree 1 to
actually work.

Thought: default to 1, but is there a way to have a PETSc option specifying
the degree you want to reduce to so it could be overridden?  Similar use
case -- maybe you've got a good preconditioner for your degree 2 case and
want to use it instead of  degree 1.

File: firedrake/preconditioners/pmg.py (hunk 6 above)
Comment (wence-) [reply firedrake/preconditioners/pmg.py:211]: It doesn't, no, measures are outside of that stuff.

File: firedrake/preconditioners/pmg.py (hunk 8 above)
Comment (wence-) [reply firedrake/preconditioners/pmg.py]: You can import those variables from the slate file and then you need to do
```python
op2.Kernel(...,
                             include_dirs=BLASLAPACK_INCLUDE.split(),
                             ldargs=BLASLAPACK_LIB.split())
```

I think

File: firedrake/preconditioners/pmg.py (hunk 9 above)
Comment (pefarrell) [reply firedrake/preconditioners/pmg.py:98]: I like Rob's idea of defaulting to 1 and having it controllable via PETSc options.

File: tests/multigrid/test_p_multigrid.py [hunk 11]
```diff
@@ -213,3 +213,65 @@ def test_p_multigrid_mixed():
...
+def test_p_fas_scalar():
+    mesh = UnitSquareMesh(4, 4, quadrilateral=True)
+    V = FunctionSpace(mesh, "CG", 7)
+    x = SpatialCoordinate(mesh)
+    u = Function(V)
+    v = TestFunction(V)
+    f = x[0]*(1-x[0]) + x[1]*(1-x[1])
+    bcs = DirichletBC(V, 0, "on_boundary")
+
+    F = inner(grad(u), grad(v))*dx - inner(f, v)*dx
```
Comment (pefarrell) [inline tests/multigrid/test_p_multigrid.py:227]: Is it possible to include a _nonlinear_ test?

File: firedrake/preconditioners/pmg.py (hunk 9 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:100]: > Unclear. DG(0) is lowest order, but some DG methods require degree 1 to actually work. Thought: default to 1, but is there a way to have a PETSc option specifying the degree you want to reduce to so it could be overridden? Similar use case -- maybe you've got a good preconditioner for your degree 2 case and want to use it instead of degree 1.

I like your idea @rckirby. One way to do this is changing `coarsen_element` to a class method and having the coarse degree as an attribute set during the initialization. I don't think there would be any negative implications if I do this. Thoughts @pefarrell ?

File: firedrake/preconditioners/low_order.py [hunk 12]
```diff
@@ -5,13 +5,17 @@
...
-    def coarsen_element(ele):
+    def coarsen_element(self, ele):
         if isinstance(ele, MixedElement) and not isinstance(ele, (VectorElement, TensorElement)):
             raise NotImplementedError("Implement this method yourself")
 
-        p = ele.degree()
-        if p == 1:
+        N = ele.degree()
+        try:
+            N = min(N)
```
Comment (pefarrell) [inline firedrake/preconditioners/low_order.py:14]: Should we check here that they're all the same degree?
Like with `N, = set(N)` that you do below.

File: firedrake/preconditioners/pmg.py [hunk 13]
```diff
@@ -179,20 +184,47 @@ def inject_state(interpolators):
...
+                except TypeError:
+                    pass
+                return N
+
+        Nf = get_max_degree(fV.ufl_element())
+        Nc = get_max_degree(cV.ufl_element())
+
+        # Coarsen the quadrature degree in a dictionary
+        # such that the ratio of GL to GLL nodes is preserved
+        def coarsen_quadrature(df, Nf, Nc):
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:212]: Happy to see the test for this below, thanks.

File: tests/multigrid/test_p_multigrid.py [hunk 14]
```diff
@@ -213,3 +209,170 @@ def test_p_multigrid_mixed():
...
+    ppc = solver.snes.getPythonContext().ppc
+    levels = ppc.getFASLevels()
+    assert levels == 3
+    assert ppc.getFASSmoother(levels-1).getLinearSolveIterations() == 0
+
+
+def test_p_fas_nonlinear_scalar():
+    mat_type = "matfree"
+    N = 4
+    dxq = dx(degree=3*N)  # here we also test coarsening of quadrature degree
```
Comment (pefarrell) [inline tests/multigrid/test_p_multigrid.py:293]: Is there any way to assert in the test that the quadrature degree was indeed modified on the coarse problem?

**Review by pefarrell — CHANGES_REQUESTED** [review]

File: tests/multigrid/test_p_multigrid.py (hunk 14 above)
Comment (pbrubeck) [reply tests/multigrid/test_p_multigrid.py:293]: There must be a way. I could try to extract the forms from the coarse SNES.

File: firedrake/preconditioners/low_order.py (hunk 12 above)
Comment (pbrubeck) [reply firedrake/preconditioners/low_order.py:14]: Sure, this can be done here as well, since there is not built-in support for different degrees.

**Review by pefarrell — DISMISSED** [review]

**Review by wence- — COMMENTED** [review]
Some comments

File: firedrake/mg/utils.py [hunk 15]
```diff
@@ -109,8 +109,6 @@ def coarse_cell_to_fine_node_map(Vc, Vf):
         assert Vc.extruded == Vf.extruded
         if Vc.mesh().variable_layers or Vf.mesh().variable_layers:
             raise NotImplementedError("Not implemented for variable layers, sorry")
-        if Vc.extruded and Vc.mesh().layers != Vf.mesh().layers:
-            raise ValueError("Coarse and fine meshes must have same number of layers")
```
Comment (wence-) [inline firedrake/mg/utils.py:113]: This has not been a restriction for a while?

File: firedrake/functionspaceimpl.py [hunk 16]
```diff
@@ -287,7 +287,7 @@ def __init__(self, mesh, element, name=None, real_tensorproduct=False):
         if type(element) is ufl.TensorElement:
             # UFL enforces value_shape of the subelement to be empty
             # on a TensorElement.
-            self.shape = element.value_shape()
+            self.shape = element.reference_value_shape()
```
[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/functionspaceimpl.py:290]: So I just quickly tested this, and saw that the symmetric tensor is storing twice the value of the off-diagonal components. Is this supposed to happen? Also, in which file should I include the test?

[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/functionspaceimpl.py:290]: The doubling in the off-diagonal happens when assembling. I guess it makes sense if you do inner(TestFunction(V), symmetric_thing)*dx. But If when one does the projection this factor of two goes away.

[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/functionspaceimpl.py:290]: The output seems alright in paraview, e.g. I define a 3-by-3 tensor on a 2D mesh and the fields look like they are supposed to. Moreover, I apply a calculator filter with sum_{ij} (a_{ij} - a{ji})^2 and get a zero.

File: firedrake/mg/embedded.py [hunk 17]
```diff
@@ -28,6 +28,9 @@ def __init__(self, element):
             degree = element.degree()
             family = lambda c: "DG" if c.is_simplex() else "DQ"
             if isinstance(cell, ufl.TensorProductCell):
+                if type(degree) is int:
+                    degree = [degree] * len(cell.sub_cells())
+
                 scalar_element = ufl.TensorProductElement(*(ufl.FiniteElement(family(c), cell=c, degree=d)
                                                             for (c, d) in zip(cell.sub_cells(), degree)))
```
[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/mg/embedded.py:35]: This code will produce the embedding DG element in `element.cell()` with `element.degree()`, regardless of `element.family()`. DG is a TPE if the cell is also TP. 
Serendipity can only be specified with an integer degree, so it might be better to construct DG via `TensorProductElement` if the degree is a tuple with different values, and we can use the standard `FiniteElement` constructor otherwise . 

Am I missing something? @rckirby any ideas?

File: firedrake/mg/utils.py (hunk 15 above)
Comment (rckirby) [reply firedrake/mg/utils.py:113]: When I was running the code with serendipity, it caught this exception.  Removing the exception worked without issue.  So probably we can keep that restriction out?

Comment (pbrubeck) [reply firedrake/mg/utils.py:113]: These hack has been properly addressed in PR #2012

File: firedrake/preconditioners/pmg.py [hunk 18]
```diff
@@ -95,124 +128,147 @@ def initialize(self, pc):
...
+                         for f in replace(form, replace_d).integrals()]) if isinstance(form, Form) else form
+
+        def coarsen_bcs(fbcs):
+            cbcs = []
+            for bc in fbcs:
+                cV_ = cV
+                for index in bc._indices:
+                    cV_ = cV_.sub(index)
+                cbc_value = self.coarsen_bc_value(bc, cV_)
+                cbcs.append(firedrake.DirichletBC(cV_, cbc_value,
```
Comment (wence-) [inline firedrake/preconditioners/pmg.py:213]: `type(bc)(cV_, ...)`

I think maybe a bigger change (open an issue to do it) is to provide `reconstruct` methods for BC objects so that you don't have to dispatch based on type when doing this stuff, but not sure.

File: firedrake/preconditioners/pmg.py [hunk 19]
```diff
@@ -95,124 +128,147 @@ def initialize(self, pc):
...
-        cctx = _SNESContext(cproblem, fctx.mat_type, fctx.pmat_type,
-                            appctx=fctx.appctx,
-                            pre_jacobian_callback=fctx._pre_jacobian_callback,
-                            pre_function_callback=fctx._pre_function_callback,
-                            post_jacobian_callback=fctx._post_jacobian_callback,
-                            post_function_callback=fctx._post_function_callback,
-                            options_prefix=fctx.options_prefix,
-                            transfer_manager=fctx.transfer_manager)
+        # FIXME setting up the _fine attribute triggers gmg injection.
+        # cctx._fine = fctx
```
Comment (wence-) [inline firedrake/preconditioners/pmg.py:262]: This comes from the workaround in `_SNESContext.compute_operators` for a PETSc bug.

File: firedrake/preconditioners/pmg.py [hunk 20]
```diff
@@ -529,51 +736,53 @@ def make_blas_kernels(self, Vf, Vc):
...
+        prolong_kernel = op2.Kernel(prolong_code, "prolongation", include_dirs=BLASLAPACK_INCLUDE.split(), ldargs=BLASLAPACK_LIB.split())
+        restrict_kernel = op2.Kernel(restrict_code, "restriction", include_dirs=BLASLAPACK_INCLUDE.split(), ldargs=BLASLAPACK_LIB.split())
+        return prolong_kernel, restrict_kernel
 
     @staticmethod
-    def get_nodes_1d(V):
-        # Return GLL nodes if V==CG or GL nodes if V==DG
-        from FIAT import quadrature
-        from FIAT.reference_element import DefaultLine
+    def get_line_element(V):
```
Comment (wence-) [inline firedrake/preconditioners/pmg.py:746]: This looks like it's the same as the free function `get_line_element` above.

File: firedrake/preconditioners/pmg.py [hunk 21]
```diff
@@ -529,51 +736,53 @@ def make_blas_kernels(self, Vf, Vc):
...
-        lam = 1.0E0 / np.prod(temp, axis=1)  # barycentric weights
-        J = np.subtract.outer(xdst, xsrc)
-        idx = np.argwhere(np.isclose(J, 0.0E0, 1E-14))
-        J[idx[:, 0], idx[:, 1]] = 1.0E0
-        J = lam/J
-        J[idx[:, 0], :] = 0.0E0
-        J[idx[:, 0], idx[:, 1]] = 1.0E0
-        J *= (1/np.sum(J, axis=1))[:, None]
-        return J
+    def get_line_nodes(V):
```
Comment (wence-) [inline firedrake/preconditioners/pmg.py:769]: I think this is also repeated from above.

**Review by wence- — COMMENTED** [review]
Modulo the relatively minor fixes, I think this is good.

File: firedrake/preconditioners/pmg.py [hunk 22]
```diff
@@ -59,15 +60,39 @@ def coarsen_element(ele):
...
-            raise ValueError
-        elif degree == 1:
+        if N <= self.coarse_degree:
             raise ValueError
 
-        return ele.reconstruct(degree=degree // 2)
+        return PMGBase.reconstruct_degree(ele, max(N // 2, self.coarse_degree))
+
+    @staticmethod
+    def reconstruct_degree(ele, N):
```
[the remark this answers is not in the document]
Comment (wence-) [reply firedrake/preconditioners/pmg.py:77]: Probably, but that requires much more invasive UFL rewriting. -> Revisit some other time.

File: firedrake/preconditioners/pmg.py (hunk 20 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:746]: Right, these are duplicates. I'm deleting the static methods from above.

**Review by wence- — APPROVED** [review]

**Comment by wence-** [comment]
Thanks!
